browser/ui/webui/brave_education/education_page_handler.cc

```diff
@@ -0,0 +1,126 @@
+#include "brave/components/brave_education/common/education_content_urls.h"
+#include "brave/components/brave_vpn/common/buildflags/buildflags.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "chrome/browser/profiles/profile.h"
+
+#if BUILDFLAG(ENABLE_BRAVE_VPN)
+#include "brave/browser/brave_vpn/vpn_utils.h"
+#endif
+
+namespace brave_education {
+
+namespace {
+
+bool IsCommandSupportedForContentType(mojom::Command command,
+                                      EducationContentType content_type) {
+  switch (content_type) {
+    case EducationContentType::kGettingStarted: {
+      constexpr auto kCommands = std::to_array(
+          {mojom::Command::kOpenRewardsOnboarding,
+           mojom::Command::kOpenWalletOnboarding,
+           mojom::Command::kOpenVPNOnboarding, mojom::Command::kOpenAIChat});
+      return base::Contains(kCommands, command);
+    }
+  }
+}
+
+bool CanShowWalletOnboarding(Profile* profile) {
+  return brave_wallet::BraveWalletServiceFactory::GetServiceForContext(profile);
+}
+
+bool CanShowRewardsOnboarding(Profile* profile) {
+  return brave_rewards::RewardsServiceFactory::GetForProfile(profile);
+}
+
+bool CanShowVPNBubble(Profile* profile) {
+#if BUILDFLAG(ENABLE_BRAVE_VPN)
+  return brave_vpn::IsAllowedForContext(profile);
+#else
+  return false;
+#endif
+}
+
+bool CanShowAIChat(Profile* profile) {
+  return ai_chat::IsAllowedForContext(profile);
+}
+
+}  // namespace
+
+EducationPageHandler::EducationPageHandler(
+    mojo::PendingReceiver<mojom::EducationPageHandler> receiver,
+    Profile* profile,
+    std::optional<EducationContentType> content_type,
+    std::unique_ptr<Delegate> delegate)
+    : receiver_(this, std::move(receiver)),
+      profile_(profile),
+      content_type_(content_type),
+      delegate_(std::move(delegate)) {}
+
+EducationPageHandler::~EducationPageHandler() = default;
+
+void EducationPageHandler::GetServerUrl(GetServerUrlCallback callback) {
+  if (!content_type_) {
+    std::move(callback).Run("");
+    return;
+  }
+  auto server_url = GetEducationContentServerURL(*content_type_);
+  std::move(callback).Run(server_url.spec());
+}
+
+void EducationPageHandler::ExecuteCommand(mojom::Command command,
+                                          ExecuteCommandCallback callback) {
+  if (!CanExecute(command)) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  switch (command) {
+    case mojom::Command::kOpenWalletOnboarding:
+      delegate_->OpenURL(GURL(kBraveUIWalletURL),
+                         WindowOpenDisposition::NEW_FOREGROUND_TAB);
+      break;
+    case mojom::Command::kOpenRewardsOnboarding:
+      delegate_->OpenRewardsPanel();
+      break;
+    case mojom::Command::kOpenVPNOnboarding:
+      delegate_->OpenVPNPanel();
+      break;
+    case mojom::Command::kOpenAIChat:
+      delegate_->OpenAIChat();
+      break;
+  }
+
+  std::move(callback).Run(true);
+}
+
+bool EducationPageHandler::CanExecute(mojom::Command command) {
```

Comment:
I definitely do not think this should be specific to brave education, it should be a generic handler like upstream so everything is in one place where we can more easily manage and control what gets added

Reply:
Everything will be right here.

The design I'm aiming for here is that education content will never be loaded on a different WebUI, and we should reject any requests for `brave.com` education content to appear in other WebUIs. `EducationPageUI` is flexible in that it can load different server URLs, depending on the `chrome://` URL. In this initial version, we just have:

```
brave://getting-started => https://brave.com/get-started
```

But in the future we could have

```
brave://getting-started/subpage => https://brave.com/get-started/subpage
```

Or even:

```
brave://whats-new => https://brave.com/whats-new
```

`EducationPageHandler` determines which commands are supported for each education page type. See the `IsCommandSupportedForPageType` function, for example. In any case, `EducationPageHandler` is responsible for all command logic.

Reply:
@bridiver can you provide an example of this?
>  it should be a generic handler like upstream so everything is in one place where we can more easily manage and control what gets added

`IsCommandSupportedForPageType` is basically controlling what commands the page including this handler can execute. The logic here in `CanExecute` is going a step further to see if this command can be ran. For example: is the feature required by the command enabled?

Are you asking for a generic (not education related) command execution handler? If there was another example of what might use a mechanism like this I guess it would be easier for me to picture. If something (??) in the future needs to do something like this too, this could potentially be broken out at that time IMO (versus trying to anticipate future usage).

Reply:
Circling back on this - I just remembered there being a somewhat generic [browser command handler](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/webui/browser_command/browser_command_handler.h) which I think is what you were mentioning.

It looks like some things (ex: what's new) have shifted logic over to [components/user_education/webui](https://source.chromium.org/chromium/chromium/src/+/main:components/user_education/webui/README.md) which has page specific type of logic (ex: see [whats_new_registry.cc](https://source.chromium.org/chromium/chromium/src/+/main:components/user_education/webui/whats_new_registry.cc)) which then seems to call the browser commands from `browser_command_handler.h`.

Given upstream has a lot more examples at the moment (and more generic things - like opening password manager, etc), I think it's fine having the logic on this page.

If we have another educational type of page (or other examples where WebUI needs to execute commands), we could look at mirroring upstream closer (have a page specific registry, migrate the commands to a common browser command handler). But we don't have any other plans at the moment. Having everything here in one module makes it easier to remove if this educational feature doesn't help w/ retention and we decide to rip it out later.